### ISSUE TITLE
xgboost generates incorrect models depending on amount of data nd free memory

### ISSUE THREAD

**Issue by KOLANICH** [dmlc/xgboost#4019]
The situation:
Let we have a large `pandas.DataFrame`, and we have already tuned hyperparams. We `sample` that DataFrame and train and save the model. Beginning from some amount of lines the resulting models start returning `nan`s.

**Comment by trivialfis** [comment]
@KOLANICH I don't really understand the procedure you described.  Can you show me a code snippet?

**Comment by KOLANICH** [comment]
>Can you show me a code snippet?

No, I have not modularized it into a snippet yet. And I'm sure that it is useless: the issue depends on amount of free RAM. Less free RAM - less the amount of rows resulting into a correct model. The lib and app doesn't crash, but the models trained are degenerate: trees in them are stumps returning `nan`.

**Comment by trivialfis** [comment]
@KOLANICH I'm not convinced that this is related to free ram.  There's no RAM checking operations in XGBoost so everything is handled by your OS's kernel and libc, and on modern OS this is not visible to user application, not even using SWAP area.

My speculation is on your particular dataset, which might be causing overflow.

**Comment by KOLANICH** [comment]
>There's no RAM checking operation in XGBoost so everything is handled by your OS's kernel and libc. On modern OS this is not visible to user application, not even using SWAP area.

I mean if it can be somewhere allocattor throws, but the exception is caught.

>My speculation is on your particular dataset, which might be causing overflow.

Is not excluded.

**Comment by trivialfis** [comment]
> but the exception is caught
Highly unlikely.  We don't catch `bad_alloc`.  There are two places that catch `std::exception` but both of them throw `dmlc::Error` immediately.  Grepping the source code should reveal it.

**Comment by KOLANICH** [comment]
OK.
